Hilt migration experiment

irfano opened the pull request:
This PR adds Hilt infrastructure. This doesn't change any user-facing feature. Details explained here: pbMoDN-35Z-p2

- WordPress class is converted to an abstract class to be used by real application and test application commonly. 
- `@AndroidEntryPoint` could be added to all fragment, activity classes but we didn't want to touch so many classes. So, `AppComponent` is converted to an EntryPoint aggregator. With this change, fragments can still access `AppComponent` class and use its `inject()` methods.
- A separate AppInitializer class is created to handle initializing in different way for real and test applications.

Hilt and dagger can be used together in the project with this PR.

### **To test:**

This change is a big one and there is no feature change.  All build variants, debug and release builds must be tested.

## Regression Notes
1. Potential unintended areas of impact
This PR intents to have no feature change. This is just infrastructure change. But since dagger is being used on every class, that means all screens can be impacted. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
I smoke tested the app, it's working as expected. For release build, I relied on automated tests.
3. What automated tests I added (or what prevented me from doing so)
It's not added any new test because there is no new feature. Tests were broken after adding Hilt, I fixed them.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

irfano committed:
Add Hilt dependency

irfano committed:
Update Dagger modules to be installed to Hilt

- Updated Dagger modules with @InstallIn(SingletonComponent::class) annotation.
- AppComponent is converted to entry point aggregator. This is a temporary change. AppComponent will be removed when Hilt migration is done.
- Added @ApplicationContext to context parameters in modules.

irfano committed:
Add initializers to start with Hilt injections

Before this commit, app was crashing at start. Because some initializing stuff were working after Hilt injections. Now they are starting together with other injections.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/WordPress.java:
```diff
@@ -168,6 +165,12 @@
+    // at application startup.
+    @Inject WebViewDataDirectorySuffixInitializer mWebViewDataDirectorySuffixInitializer;
+    // Well sql is initialized with `WPWellSqlConfig` for debug, `WellSqlConfig` for release.
+    @Inject WellSqlInitializer mWellSqlInitializer;
```
Thank you for adding these helpful comments!

renanferrari commented on WordPress/src/androidTest/java/org/wordpress/android/WordPressTestRunner.kt:
```diff
@@ -5,10 +5,10 @@ import android.content.Context
 class WordPressTestRunner : AndroidJUnitRunner() {
     override fun newApplication(classLoader: ClassLoader, className: String, context: Context): Application {
-        return super.newApplication(classLoader, WordPressTest::class.java.name, context)
+        return super.newApplication(classLoader, WordPress::class.java.name, context)
```
I might be missing something here, but since we're not using a custom application for instrumentation tests anymore, do we still need to use this custom test runner?

irfano replied:
You're right to catch here. This line doesn't make sense. But I didn't want to focus test in this commit.
My next commit will be for fixing test, and I'm creating a new test application. I'll explain the details for that commit.

renanferrari reviewed (commented):
Thank you for the changes, @irfano! They are looking pretty good so far. Just left one minor question below.

I've also submitted #15659 to force trigger CI here, and it seems that we have some failing lint checks when running `./gradlew ktlint`. Could you take a look at it?

irfano committed:
Correct typo to fix ktlint error

irfano commented:
> Thank you for the changes, @irfano! They are looking pretty good so far. Just left one minor question below.
> 
> I've also submitted #15659 to force trigger CI here, and it seems that we have some failing lint checks when running `./gradlew ktlint`. Could you take a look at it?

Thanks! I fixed ktlint error.

irfano committed:
Update Hilt version

irfano committed:
Move initialization logic from WordPress to AppInitializer

Hilt test requires test application but test applications have some drawbacks. While real applications can inject before onCreate, test applications can only inject after onCreate. Thus, AppInitializer is created and will be used as a rule in test cases.

irfano commented:
Hilt is working for real application and unit tests, but UI tests require architecture update. 

Hilt test requires test application but test applications have some drawbacks. While real applications can inject before onCreate, test applications can only inject after onCreate. 

Because of test application's injection order problem, [AppInitializer](https://github.com/wordpress-mobile/WordPress-Android/blob/be0ed5da81f090c02dd5f56eb9dfe5bf269b4096/WordPress/src/main/java/org/wordpress/android/AppInitializer.kt) is created in be0ed5d and will be used as [InitializationRule](https://github.com/wordpress-mobile/WordPress-Android/blob/be0ed5da81f090c02dd5f56eb9dfe5bf269b4096/WordPress/src/androidTest/java/org/wordpress/android/InitializationRule.kt) in test cases.

[Wordpress](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/WordPress.java) has many dependency to other classes. It has: 
12 public static functions, 
11 public static variables,
6 public functions,
2 public variables.
These members are being used by other classes. They aren't removed on purpose since other classes are using them. Even some logics moved to [AppInitializer](https://github.com/wordpress-mobile/WordPress-Android/blob/be0ed5da81f090c02dd5f56eb9dfe5bf269b4096/WordPress/src/main/java/org/wordpress/android/AppInitializer.kt), gateway functions were left in WordPress. 

Important details for be0ed5d:
- New structure is tested on [LoginTest](https://github.com/wordpress-mobile/WordPress-Android/blob/be0ed5da81f090c02dd5f56eb9dfe5bf269b4096/WordPress/src/androidTest/java/org/wordpress/android/e2e/LoginTests.java). Other test use cases will be fixed in future commits.
- Before, WorkManager config difference for debug and release builds were being handled in [WordPressDebug](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/debug/java/org/wordpress/android/WordPressDebug.java). Now it's handled with `BuildConfig.DEBUG` in `AppInitializer`.
- Lazy initialization were being used for some variables in WordPress. Kotlin's lazy property feature is used to provide same feature. (`restClientUtils`, `restClientUtilsV1_1`, ..., `defaultUserAgent`, `userAgent`)

irfano committed:
Fix build failure caused by int casting problem on AppInitializer

renanferrari reviewed (commented):
@irfano Thank you for the changes! 

I spent some time smoke testing the app, specifically features that I know were affected by the introduction of `AppInitializer`, and everything seemed to be working as expected. There's always the possibility for something to be silently failing, but I think we're good for now.

As for the changes you made to the code, they all make sense to me so far. As you mentioned before, I think the main hurdle right now is to figure out how to handle those classes that are casting `Application` to `WordPress`. How are you planning to tackle that?

>Even some logics moved to AppInitializer, gateway functions were left in WordPress.

>Before, WorkManager config difference for debug and release builds were being handled in WordPressDebug. Now it's handled with BuildConfig.DEBUG in AppInitializer.

>Lazy initialization were being used for some variables in WordPress. Kotlin's lazy property feature is used to provide same feature

Good calls! 👍

---

Aside from the failing UI tests, Lint is also failing with the following error:

`Class referenced in the manifest, .WordPressRelease, was not found in the project or the libraries`

I haven't looked into this. Do you have any idea what could be causing it?

irfano committed:
Fix unresolved class error in AndroidManifest by adding package attribute

irfano commented:
Thanks for your review @renanferrari !

> how to handle those classes that are casting Application to WordPress. How are you planning to tackle that?

I think the only way is changing the code where accessing [WordPress](https://github.com/wordpress-mobile/WordPress-Android/blob/07a3995683fddaa0df8515eb77340b2ea27de82a/WordPress/src/main/java/org/wordpress/android/WordPress.java). I plan to change them to access [AppInitializer](https://github.com/wordpress-mobile/WordPress-Android/blob/07a3995683fddaa0df8515eb77340b2ea27de82a/WordPress/src/main/java/org/wordpress/android/AppInitializer.kt) instead of [WordPress](https://github.com/wordpress-mobile/WordPress-Android/blob/07a3995683fddaa0df8515eb77340b2ea27de82a/WordPress/src/main/java/org/wordpress/android/WordPress.java). After that change, I probably will remove public gateway functions in [WordPress](https://github.com/wordpress-mobile/WordPress-Android/blob/07a3995683fddaa0df8515eb77340b2ea27de82a/WordPress/src/main/java/org/wordpress/android/WordPress.java).
(Another solution would be using WordPress as a base Application, but that was our main problem hor Hilt testing, we can't do that.)

---
> Lint is also failing with the following error:
> `Class referenced in the manifest, .WordPressRelease, was not found in the project or the libraries`

Fixed that, it was a small mistake. I'd left release build untested. But then I realized I can test release build by just changing signing config with debug signing, at least I can be sure building is not failing. I did it, and now it's working.

hichamboushaba commented on WordPress/src/androidTest/java/org/wordpress/android/WordPressTest.kt:
```diff
@@ -1,11 +1,38 @@
+        super.onCreate()
+    }
+
+    override fun androidInjector(): AndroidInjector<Any> {
+        if (!this::injector.isInitialized) {
+            injector = EntryPoints.get(
+                    applicationContext,
+                    AndroidInjectorEntryPoint::class.java
+            ).injector()
+        }
+        return injector
```
@irfano I did the same mistake here on my initial migration in WCAndroid, but later one we found out that when running multiple tests on the same session, the caching of the `AndroidInjector` causes an issue of having inconsistent copies of dependencies.
For more details check my last PR in WCAndroid: https://github.com/woocommerce/woocommerce-android/pull/5445

irfano replied:
Thank you for warning me. I'll consider it while fixing UI tests. 🙏🏻

hichamboushaba commented on WordPress/src/main/java/org/wordpress/android/AppInitializer.kt:
```diff
@@ -0,0 +1,981 @@
+    // at application startup.
+    @Inject lateinit var mWebViewDataDirectorySuffixInitializer: WebViewDataDirectorySuffixInitializer
+
+    // Well sql is initialized with `WPWellSqlConfig` for debug, `WellSqlConfig` for release.
+    @Inject lateinit var mWellSqlInitializer: WellSqlInitializer
```
np, depending on the comment here to ensure WellSql is initialized is not the best approach IMO, since you added a separate class for the initialization, I would suggest adding an explicit `init` function to it and calling it from the `AppInitializer#init` function, instead of depending on the `init` bloc.
This would also ensure that the behavior is correct even if Dagger's behavior changed from eager injecting to lazier injecting here 🙂.

Same remark for `WebViewDataDirectorySuffixInitializer`.

irfano replied:
I initialized them with `@Inject` annotation instead of `AppInitializer#init` function, because other `@Inject` initializations need them.

If I init `WellSql` in `AppInitializer#init`, `AccountStore` crashes because of uninitialized `WellSql`.

I want to initialize `WebViewDataDirectorySuffixInitializer` and `WellSqlInitializer` before everything else. I think doing it in `AppInitializer#init` is not a right way to do that but we can discuss alternative ways.

What is your main concern here? Does it seem weird to inject a class just for initializing them?

hichamboushaba replied:
> If I init WellSql in AppInitializer#init, AccountStore crashes because of uninitialized WellSql.

Yes, I forgot about this, so maybe calling `init` directly in the app's `onCreate` 🤔 

> What is your main concern here? Does it seem weird to inject a class just for initializing them?

It's nothing major, but just two small issues IMO:
1. It's not super clear, and that's why we depend on the comment to explain why it's needed.
2. It depends on Dagger's logic, if Dagger was to inject the dependencies lazily by default (like Koin for example), this wouldn't work, since the class won't be instantiated until it's injected.

Anyway, it's just a small point, so whatever you and @renanferrari agree on, for WCAndroid, we didn't create any class or WellSql's initialization, and instead it's done manually in the Application's `onCreate`.

irfano replied:
> Yes, I forgot about this, so maybe calling init directly in the app's onCreate 🤔

It's called in [WooCommerce](https://github.com/woocommerce/woocommerce-android/blob/24dd260a478e55c6905e99bdd1012052ec655592/WooCommerce/src/main/kotlin/com/woocommerce/android/WooCommerce.kt#L27) in WCAndroid. At first I thought doing the same, but then I thought keeping application class minimal could be a better approach. Otherwise, I'd have to handle `WellSql` initialization also in test application, like [this](https://github.com/woocommerce/woocommerce-android/blob/0a85b39af97b2232000ca33fbd2e22b4f59ecabf/WooCommerce/src/androidTest/kotlin/com/woocommerce/android/WooCommerceTest.kt#L27) made in WCAndroid.

What about initialize them in `AppInitializer`'s init and injecting `AccountStore` lazily?
https://gist.github.com/irfano/790096d16ed4b0c4ca1adc366102d191

hichamboushaba replied:
> What about initialize them in AppInitializer's init and injecting AccountStore lazily?

Yes, this is another possibility, I like it 👍

hichamboushaba reviewed (commented):
Good job @irfano, I left some comments on the code, but nothing major.

But UI tests are not working currently, they sometimes crash with the following exception:
```
Caused by: java.lang.ClassCastException: org.wordpress.android.WordPressTest_Application cannot be cast to org.wordpress.android.WordPress
        at org.wordpress.android.push.GCMMessageService.onCreate(GCMMessageService.java:41)Caused by: java.lang.ClassCastException: org.wordpress.android.WordPressTest_Application cannot be cast to org.wordpress.android.WordPress
        at org.wordpress.android.push.GCMMessageService.onCreate(GCMMessageService.java:41)
```

And sometimes they just don't start, and they show a toast about DB inconsistency.

irfano commented:
Thanks for your review @hichamboushaba 
I'm aware of failing UI tests and currently working on it.

irfano committed:
Init WellSql and WebViewDataDirectorySuffix in init block of AppInitializer

irfano committed:
Fix UserAgentTest errors

- Added HiltAndroidTest to UserAgentTest
- Moved getUserAgent(), getDefaultUserAgent(), USER_AGENT_APPNAME from WordPress to AppInitializer

irfano commented on WordPress/src/androidTest/java/org/wordpress/android/InitializationRule.kt:
```diff
@@ -23,17 +23,27 @@ class InitializationRule : TestRule {
+        // InitializationRule must be initialized only once. Otherwise some static functions in AppInitializer throws
+        // exception, if they are initialized more than once. (e.g. WebView.setDataDirectorySuffix(),
+        // EventBusBuilder.installDefaultEventBus)
+        private var isInitialized = false
```
With `@Rule` annotation, the TestRule runs for each test cases. That means `InitializationRule` will be called for each test cases. But that causes crash because some functions in `AppInitializer` must be called only once for application lifecycle. 

With `isInitialized`, I ensured that `AppInitializer` will be called once. The disadvantage of this way can be that same `AppInitializer` instance will be used for each test cases. If I encounter this problem, I'll try to clean resources in `AppInitializer` before test cases.

I tried to use `@ClassRule` to run test rule once for this class, but it didn't work because of Hilt's Entry Point. I couldn't find the reason. I may try to use `@ClassRule` again later.

irfano committed:
Merge branch 'develop' into feature/hilt-migration

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/modules/AppComponent.java
#	build.gradle

irfano committed:
Remove caching instances for UI tests

Since SingletonComponent is recreated for each test cases, caching AndroidInjector and AppInitializer causes failing UI tests. AndroidInjector and AppInitializer are no longer cached for UI tests. But some static functions were crashing while AppInitializer is recreating. Added a boolean to AppInitializer to manage reinitializing of these functions.

irfano committed:
Make WordPress abstract to be used by real and test applications

Other classes are casting application to WordPress but WordPress couldn't be used in UI tests. And test application (WordPressTest) was crashing.
With this common abstract WordPress application, other classes can cast application to WordPress safely.

irfano committed:
Add DummyInterceptorModule for UI tests

irfano commented on WordPress/src/androidTest/java/org/wordpress/android/InitializationRule.kt:
```diff
@@ -14,36 +14,23 @@ class InitializationRule : TestRule {
-        // InitializationRule must be initialized only once. Otherwise some static functions in AppInitializer throws
-        // exception, if they are initialized more than once. (e.g. WebView.setDataDirectorySuffix(),
-        // EventBusBuilder.installDefaultEventBus)
-        private var isInitialized = false
```
I removed `isInitialized` here. This was a wrong approach. `AppInitializer` should be recreated for each test cases.

irfano committed:
Fix UI tests by implementing Hilt test architecture

irfano commented on WordPress/src/androidTest/java/org/wordpress/android/modules/DummyInterceptorModule.kt:
```diff
@@ -0,0 +1,19 @@
+import okhttp3.Interceptor
+import javax.inject.Named
+
+@TestInstallIn(
+        components = [SingletonComponent::class],
+        replaces = [InterceptorModule::class]
+)
+@Module
+class DummyInterceptorModule {
+    @Provides @IntoSet @Named("network-interceptors")
+    fun provideNetworkInterceptor(): Interceptor = Interceptor { it.proceed(it.request()) }
+}
```
@hichamboushaba I copied `DummyInterceptorModule` from WCAndroid. It works but I couldn't understand what this actually does. Is `@Named("network-interceptors")` required for fluxc library? Should I check fluxc documents?

hichamboushaba replied:
@irfano I need to check, but if I remember correctly the reason we needed this it was just to replace the actual `InterceptorModule` in tests with a dummy one, so the main thing is:
```kotlin
@TestInstallIn(
        components = [SingletonComponent::class],
        replaces = [InterceptorModule::class]
)
```
and not the interceptor itself.
I didn't test this, but maybe you can just keep the module empty, without the provided interceptor at all 🤔

irfano replied:
Thanks for the explanation.

I learnt that "network-interceptors" is mainly for logging, "interceptors" is for mocking [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1155).
WCAndroid is using [Flipper](https://fbflipper.com/), WPAndroid is using [Stetho](http://facebook.github.io/stetho/) for network debugging. By adding DummyInterceptorModule, WCAndroid removed Flipper for testing. I couldn't understand the exact reason why do we need to cancel network debugging, but I will also do the same for WPAndroid.

I mean I removed Stetho for testing by adding DummyInterceptorModule.

hichamboushaba replied:
We remove the logging because it adds an overhead which is not necessary for tests, since the network requests are already mocked, but I don't think it will hurt if it stayed, if it's just pure logging.

irfano replied:
It makes sense now. Thanks! 👍🏻

irfano committed:
Fix SignUpTests

irfano commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/flows/SignupFlow.java:
```diff
@@ -87,5 +88,6 @@ public SignupFlow dismissInterstitial() {
     public void confirmSignup() {
         // Confirm signup
         waitForElementToBeDisplayed(R.id.nav_sites);
+        new MySitesPage().go();
     }
```
Before this commit, the test was passing successfully. Then activity was finishing while recylerview is still refreshing and then [ReaderExpandableTagsView](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderExpandableTagsView.kt) on "Reader" screen was crashing because of unaccessible Hilt objects.

If test was waiting enough for reloading of recyclerview, test was passing. But sometimes it was failing.

Waiting for 2 seconds at `confirmSignup()` can prevent failing.
I chose navigating "My Sites" screen that also prevent failing.

renanferrari replied:
Just replicating here what I said on Slack, for visibility: I personally think both approaches are fine in this case, but maybe we should add an extra comment there, because it’s not immediately obvious that removing that piece of code could cause the test to fail.

renanferrari reviewed (commented):
Thank you for the changes, @irfano! All CI checks are now passing, including UI tests 🎉 I also smoke tested the app and, from what I can tell, it's still working as expected.

The code is looking good overall and I like the approach used to solve the `WordPress` casting issue 👍

@hichamboushaba Thank you for your assistance with this so far! Let us know if you have any additional comments.

irfano committed:
Add comment to confirmSignup about why navigating My Site is added to the test

irfano commented:
This PR is canceled and the migration is being implemented in #16143.
